lib/ansible/module_utils/ipa.py

```diff
@@ -0,0 +1,102 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+try:
+    import json
+except ImportError:
+    import simplejson as json
+
+from ansible.module_utils.pycompat24 import get_exception
+from ansible.module_utils.urls import fetch_url
+from ansible.module_utils.six.moves.urllib.parse import quote
+from ansible.module_utils.six import b, PY3
+
+class IPAClient(object):
+    def __init__(self, module, host, port, protocol):
+        self.host = host
+        self.port = port
+        self.protocol = protocol
+        self.module = module
+        self.headers = None
+
+    def get_base_url(self):
+        return '%s://%s/ipa' % (self.protocol, self.host)
+
+    def get_json_url(self):
+        return '%s/session/json' % self.get_base_url()
+
+    def login(self, username, password):
+        url = '%s/session/login_password' % self.get_base_url()
+        data = 'user=%s&password=%s' % (quote(username, safe=''), quote(password, safe=''))
+        headers = {'referer': self.get_base_url(),
+                   'Content-Type': 'application/x-www-form-urlencoded',
+                   'Accept': 'text/plain'}
+        try:
+            resp, info = fetch_url(module=self.module, url=url, data=b(data), headers=headers)
+            status_code = info['status']
+            if status_code not in [200, 201, 204]:
+                self._fail('login', info['msg'])
+
+            self.headers = {'referer': self.get_base_url(),
+                            'Content-Type': 'application/json',
+                            'Accept': 'application/json',
+                            'Cookie': resp.info().get('Set-Cookie')}
+        except Exception:
+            e = get_exception()
+            self._fail('login', str(e))
+
+    def _fail(self, msg, e):
+        if 'message' in e:
+            err_string = e.get('message')
+        else:
+            err_string = e
+        self.module.fail_json(msg='%s: %s' % (msg, err_string))
+
+    def _post_json(self, method, name, item=None):
+        if item is None:
+            item = {}
+        url = '%s/session/json' % self.get_base_url()
+        data = {'method': method, 'params': [[name], item]}
+        try:
+            resp, info = fetch_url(module=self.module, url=url, data=b(json.dumps(data)), headers=self.headers)
+            status_code = info['status']
+            if status_code not in [200, 201, 204]:
+                self._fail(method, info['msg'])
+        except Exception:
+            e = get_exception()
+            self._fail('post %s' % method, str(e))
+
+        if PY3:
+            charset = resp.headers.get_content_charset('latin-1')
+        else:
+            response_charset = resp.headers.getparam('charset')
+            charset = response_charset if response_charset else 'latin-1'
```

Comment:
ternary isn't available in python-2.4  If the freeipa module works on python2.4 then this will need to be expanded into a regular if-else.